lib/Transforms/LiquidFixpointBuilder/VariablesEnvironmentImmutable.cpp

```diff
@@ -0,0 +1,577 @@
+    {
+      blockGuard = "true";
+    }
+    else
+    {
+      std::vector<std::string> predecessors;
+      {
+	auto getPredRes = functionBlockGraph.GetPreviousBlocks(blockName, predecessors);
+	if (!getPredRes.Succeeded) { return getPredRes; }
+      }
+      
+      std::vector<std::string> predecessorTransitionGuards;
+      for (const auto& predecessor : predecessors)
+      {
+	std::string transitionGuardName = "__transition__"s + predecessor + "__"s + blockName;
+
+	// (Bug?) The recursive call later gives unallocated strings here, until we find a predecessor that is the same as the current block.
+	std::string predecessorEntryGuard;
+	{
+	  if (predecessor == blockName)
+	  {
+	    predecessorEntryGuard = transitionGuardName;
+	  }
+	  else
+	  {
+	    auto getPredEntryRes = getBlockGuard(predecessor, predecessorEntryGuard);
+	    if (!getPredEntryRes.Succeeded) { return getPredEntryRes; }
+	  }
+	}
+	
+	predecessorTransitionGuards.emplace_back("("s + predecessorEntryGuard + " && "s + transitionGuardName + ")"s);
+      }
+	
+      blockGuard = "("s + RefinementUtils::StringJoin(" || ", predecessorTransitionGuards) + ")"s;
+    }
+    
+    cachedBlockGuards[blockName] = blockGuard;
+    return ResultType::Success();
+  }
+
+  ResultType VariablesEnvironmentImmutable::addVariableToBlockAndSuccessors(const std::string& blockName, const std::string& transitionGuardName)
+  {
+    std::vector<std::string> successors;
+    {
+      auto getSuccRes = functionBlockGraph.GetAllSuccessorBlockNames(blockName, successors);
+      if (!getSuccRes.Succeeded) { return getSuccRes; }
+    };
+    
+    for (const auto& successor : successors)
+      {
+	variablesMappingsPerBlock[successor][transitionGuardName] = transitionGuardName;
+	variablesValuesPerBlock[successor].emplace(transitionGuardName);
+      }
+    
+    return ResultType::Success();
+  }
+  
+  ResultType VariablesEnvironmentImmutable::initializeBlockGuards()
+  {
+    std::vector<std::string> blockNames;
+    {
+      auto getBlockRes = functionBlockGraph.GetAllBlockNames(blockNames);
+      if (!getBlockRes.Succeeded) { return getBlockRes; }
+    }
+    
+    for (const auto& blockName : blockNames)
+    {
+      std::vector<std::string> successors;
+      {
+	auto getSuccRes = functionBlockGraph.GetSuccessorBlocks(blockName, successors);
+	if (!getSuccRes.Succeeded) { return getSuccRes; }
+      }
+      
+      for(const auto& successor : successors)
+      {
+	std::string transitionGuardName = "__transition__"s + blockName + "__"s + successor;
+	{
+	  auto createBinderRes = constraintBuilder.CreateFutureBinder(transitionGuardName, FixpointType::GetBoolType());
+	  if (!createBinderRes.Succeeded) { return createBinderRes; }
+	}
+	
+	{
+	  auto addVarRes = addVariableToBlockAndSuccessors(successor, transitionGuardName);
+	  if (!addVarRes.Succeeded) { return addVarRes; }
+	}
+      }
+    }
+    
+    for (const auto& blockName : blockNames)
```

Comment:
get rid of block guards

Reply:
This is on the next part of the `TODO` list - Replacing `blockGuards` with a purely Fixpoint approach.

Reply:
This is on the next part of the `TODO` list - Replacing `blockGuards` with a purely Fixpoint approach.